File: join_data.R
#!/usr/bin/env Rscript

# Проверка
if (!requireNamespace("dplyr", quietly = TRUE)) {
  stop("Пакет 'dplyr' не установлен.")
}

# Загрузка библиотеки
library(dplyr)

# Пути к входным файлам
input_metadata <- "data/sample_metadata.csv"
input_ms       <- "data/mass_spec_results.csv"

# Проверка существования входных файлов
if (!file.exists(input_metadata)) stop("Файл не найден: ", input_metadata)
if (!file.exists(input_ms))       stop("Файл не найден: ", input_ms)

# Чтение данных (без преобразования строк в факторы)
sample_metadata   <- read.csv(input_metadata, stringsAsFactors = FALSE)
mass_spec_results <- read.csv(input_ms,       stringsAsFactors = FALSE)

# Анти-левый джойн: строки из метаданных, отсутствующие в результатах МС
anti_left <- anti_join(
  sample_metadata,
  mass_spec_results,
  by = "sample_id"
)

# Анти-правый джойн: строки из результатов МС, отсутствующие в метаданных
anti_right <- anti_join(
  mass_spec_results,
  sample_metadata,
  by = "sample_id"
)

# Анти-внешний джойн: объединение двух предыдущих результатов с пометкой источника
anti_outer <- bind_rows(
  anti_left  %>% mutate(источник = "sample_metadata"),
  anti_right %>% mutate(источник = "mass_spec_results")
)

# Сохранение результатов в CSV (без номеров строк)
write.csv(anti_left,  "data/anti_left.csv",  row.names = FALSE)
write.csv(anti_right, "data/anti_right.csv", row.names = FALSE)
write.csv(anti_outer, "data/anti_outer.csv", row.names = FALSE)

# Уведомление об успешном завершении
cat("Выполнены. Результаты сохранены в папку data/.\n")

File: join_data.py
#!/usr/bin/env python3

import pandas as pd
import os
import sys
import logging

# Настройка логирования
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger()

INPUT_DIR = "/input"
OUTPUT_DIR = "/output"

# Убедимся, что output существует и имеет права 777
os.makedirs(OUTPUT_DIR, exist_ok=True)
os.chmod(OUTPUT_DIR, 0o777)

# Пути к файлам
metadata_path = os.path.join(INPUT_DIR, "sample_metadata.csv")
ms_path = os.path.join(INPUT_DIR, "mass_spec_results.csv")

# Проверка наличия файлов
if not os.path.exists(metadata_path):
    logger.error(f"Input file not found: {metadata_path}")
    sys.exit(1)
if not os.path.exists(ms_path):
    logger.error(f"Input file not found: {ms_path}")
    sys.exit(1)

logger.info("Loading input data...")
metadata = pd.read_csv(metadata_path)
ms = pd.read_csv(ms_path)

logger.info(f"Loaded {len(metadata)} rows from sample_metadata")
logger.info(f"Loaded {len(ms)} rows from mass_spec_results")

# Inner join
logger.info("Performing inner join...")
inner = pd.merge(metadata, ms, on="sample_id", how="inner")
inner.to_csv(os.path.join(OUTPUT_DIR, "inner_join.csv"), index=False)
logger.info(f"Inner join completed: {len(inner)} rows saved.")

# Left join (metadata as left)
logger.info("Performing left join (sample_metadata left)...")
left = pd.merge(metadata, ms, on="sample_id", how="left")
left.to_csv(os.path.join(OUTPUT_DIR, "left_join.csv"), index=False)
logger.info(f"Left join completed: {len(left)} rows saved.")

# Right join
logger.info("Performing right join (mass_spec_results right)...")
right = pd.merge(metadata, ms, on="sample_id", how="right")
right.to_csv(os.path.join(OUTPUT_DIR, "right_join.csv"), index=False)
logger.info(f"Right join completed: {len(right)} rows saved.")

# Outer join
logger.info("Performing outer join...")
outer = pd.merge(metadata, ms, on="sample_id", how="outer")
outer.to_csv(os.path.join(OUTPUT_DIR, "outer_join.csv"), index=False)
logger.info(f"Outer join completed: {len(outer)} rows saved.")

logger.info("All joins completed successfully.")
